enhancement(sinks): Use variance calculation to better adapt to RTT times in ARC

bruceg opened the pull request:
This adds a variance (the square of standard deviation) calculation to the EWMA statistic, which is then used to provide an estimate of the expected deviation of the RTT average. That is then used to provide a better upper bound for when to scale down the concurrency limit, which allows us to keep the concurrency high for more of the time.

Closes #7371

bruceg committed:
Add EWMA type with variance calculations

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Use new EwmaVar type in ARC controller (no behavior changes)

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Use the new variance in the ARC calculations

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Mark ARC config as `deny_unknown_fields`

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg commented on docs/cue/reference/components/sinks.cue:
```diff
@@ -260,11 +260,13 @@ components: sinks: [Name=string]: {
-											description: "When comparing the past RTT average to the current measurements, we ignore changes that are less than this ratio higher than the past RTT. Valid values are greater than or equal to 0. Larger values cause the algorithm to ignore larger increases in the RTT."
-											required:    false
-											type: float: default: 0.05
+										rtt_deviation_scale: {
```
I would be quite open to a better name for this configuration. The term "deviation" has too many bad connotations, but I wasn't coming up with anything better that was still descriptive.

blt replied:
I'm fine with deviation as a word, used here in a technical sense. That said, considering this it strikes me that having a measurement of the skew of the distribution would be interesting as well.

bruceg replied:
Agreed, it would be interesting, but I'm not sure how useful in the end. Also, I have no idea how to even calculate that given the EWMA scheme in use.

blt replied:
Fair point. More an idle thought than anything.

blt reviewed (approved):
Tidy work.

bruceg commented:
So the test failure appears to be… floating point subtraction weirdness. The timing of the simulated events coming in is supposed to be exactly 101ms (well, actually 100ms, long story). However, the value comes in as a number of seconds, and ends up flipping back and forth between 0.101 exactly and 0.10100000000000002. This ends up having the scaled deviation smaller than the difference between the two values, resulting in the concurrency being throttled repeatedly.

I can "fix" this by increasing the deviation scale for the self-tests, but that means we're running the tests with a different value than what we make users use. I could possibly truncate the incoming RTT values to, say, nanoseconds, but that seems like a nasty kludge. Does anybody have other thoughts where to go here?

tobz commented:
I hate to be annoying here, but: can we actually show a before/after of the ARC metrics with this change? I know it's hard to capture in unit tests, so I think a simple comparison via DD graphs could suffice.

Ideally, we'd want to be able to show like... Vector with entirely default concurrency values, and then with ARC turned on.  Long-term, it'd be great to actually have some sort of test harness where we could quantitatively show that ARC gets us within some percentage of the theoretical limit... but that's pie in the sky.

jszwedko commented:
I'd like to see the HTTP concurrency test suite we have run too for before/after: https://github.com/timberio/http_test_server/blob/master/bin/run-concurrency-test-suite

jszwedko commented:
@bruceg what would you think about adding another test for this too to the ARC tests in here that mimics the behavior we were seeing with the Datadog endpoint?

bruceg commented:
> @bruceg what would you think about adding another test for this too to the ARC tests in here that mimics the behavior we were seeing with the Datadog endpoint? Specifically one of these: https://github.com/timberio/vector/tree/master/tests/data/adaptive-concurrency

Those tests are showing significant variations from the existing parameters, so it is possible they will already reflect the improved performance. Some are showing degradation, however, so I am trying to figure out if that's a glitch or an indication of a new problem.

bruceg committed:
Work around a problem with the "constant link" test statistics

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Adjust test data for new controller

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Rework the ARC test harness to avoid the polling loop

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Adjust ARC default values

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Merge remote-tracking branch 'origin/master' into arc-ewma-variance

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg commented on tests/data/adaptive-concurrency/shallow-knee.toml:
```diff
@@ -12,12 +12,12 @@ mean = [9, 11]
 [stats.rate]
-max = [108, 108]
-mean = [87, 88]
+max = [103, 108]
+mean = [87, 89]
```
I think this is our one regression, that this one test showed a maximum rate decrease. However, the mean showed a trivial increase, so the overall performance is still up a hair.

bruceg commented:
Here is some evidence from the Datadog dashboard

Config file:
```toml
[sources.source]
  type = "file"
  data_dir = "/tmp/vector"
  include = ["/tmp/vector/*.txt"]

[sinks.datadog_logs]
  type                = "datadog_logs"
  inputs              = ["source"]
  api_key             = "b0a12175f352603d81a65ad00f0eccf7"
  compression         = "gzip"
  encoding.codec      = "json"
  healthcheck.enabled = true
  batch.size = 1000000
  request.concurrency = # SEE BELOW
  request.rate_limit_num = 1000000
```

The files were generated by @blt's [lading](https://github.com/blt/lading), configured to generate data at 100MB/sec across four files.

`concurrency = 100`
![concurrency=100](https://user-images.githubusercontent.com/477552/126586748-6d7e9e18-8632-467c-b91b-7ae514d95e82.png)

`concurrency = "adaptive"` *before* (10cf4cde9)
![before concurrency=adaptive](https://user-images.githubusercontent.com/477552/126586794-f0b1c028-d6d1-4130-8620-1d3ddbd07c3f.png)

`concurrency = "adaptive"` *after* (f870891cd)
![concurrency=adaptive after](https://user-images.githubusercontent.com/477552/126586819-e898e5af-3696-40ec-b9ec-dfba806fa922.png)

There are still some questions to be answered here, but I don't think any of them are relevant to this PR. First, why is DD reporting 3-4 times as much bytes and events ingested as we think we're sending. Second, why does the average concurrency never go above 10 even when we set the concurrency to 100? Given that limit, I suspect the increase in performance caused by this change is understated by the above graphs.

bruceg committed:
One more test data tweak

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

blt commented:
@bruceg my +1 still stands. I think the graphs here are convincing, questions about Datadog behavior not withstanding.

jszwedko commented:
I ran the external test suite we have. Some scenarios have improved, but curiously the simple 429 and 503 cases seem to have regressed.

Would you be able to take a look @bruceg ? I'm also curious if the internal tests we have should have caught this or not.

Before:

![Uploading before.png…]()

After:

![Uploading after.png…]()

bruceg commented:
Ok, that's ugly. Almost all of those scenarios show a significant regression (all except for `queue_limit` AFAICT). So we either have a situation where the simulator is producing an unrealistic enough scenario to fool the new variance calculation, or the calculation only works for a limited set of scenarios like experienced with the Datadog logs ingest service. In either case, this needs more investigation.

bruceg committed:
Calculate `reached_limit` *after* starting a request

Previously, the ARC controller determined if the concurrency limit had
been reached before starting a request. In this way, the controller
would only receive an indication that the limit was reached if the sink
tried to send at least one more than the limit. This negatively affects
performance, as the controller will only increase the limit if it was
reached.

This change calculates the value after starting a request, which is when
the indication of how many requests are in flight is changed. This also
drops a mutex lock from the `acquire` code path.

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Update tests parameters for algorithm tweak

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Merge remote-tracking branch 'origin/master' into arc-ewma-variance

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg commented:
With the revised changes to how `reached_limit` is calculated, the algorithm now responds much better in the benchmarks. Unfortunately, there is still an oddity in some of the tests, notably the "added load" and "removed load" tests (before change is on top, after change is below):

![natural_added_load](https://user-images.githubusercontent.com/477552/128237050-2fca9e86-6ef5-4d4e-9302-c69706b0c4c1.png)
Notice that the concurrency stays high, but the request rate drops to zero briefly. I *think* what is happening is that the sudden doubling in response latency causes a large number of requests to stay outstanding for too long. As seen, the concurrency starts dropping after half way, so I think this is working as good as can be expected, it just looks ugly.

![natural_removed_load](https://user-images.githubusercontent.com/477552/128237550-1128eb30-4500-469f-9de4-34cb199c4b6f.png)
This one is more puzzling. You can see the concurrency starting to plateau just before half way, and then rise again after the RTT drops at the half way point. However, the request rate stays low. There is nothing in my changes that should cause that change, so I have no explanation here.

![natural_medium_variance](https://user-images.githubusercontent.com/477552/128237899-d05a9672-fedd-437f-ba2b-b0743237f961.png)
Similarly, in this case, the new algorithm reaches a stable concurrency nearly double what it was before these changes, but the request rate is lower.

Does anybody have ideas why I might be seeing lower request rates with *higher* concurrency levels?

jszwedko commented:
> Similarly, in this case, the new algorithm reaches a stable concurrency nearly double what it was before these changes, but the request rate is lower.
> 
> Does anybody have ideas why I might be seeing lower request rates with _higher_ concurrency levels?

For this one, I think this is because that test has each request take longer the more concurrent requests there are. At 21 concurrent requests, which seems to be the peak there for "after changes", each request should take 2353 ms for a request rate of about 9 requests / s which matches up with the plot. The "before changes" was able to achieve higher throughput with less number of concurrent requests because each one was shorter. What is interesting to me is that the algorithm decided to increase the concurrent requests even higher which should result in even less requests / second. I'll try running that one for longer than 60s to see what happens.

> This one is more puzzling. You can see the concurrency starting to plateau just before half way, and then rise again after the RTT drops at the half way point. However, the request rate stays low. There is nothing in my changes that should cause that change, so I have no explanation here.

I am a bit concerned about how it reacted there. It seems to have not correctly increased concurrency in absence of removed load. Could it be taking a while to readjust to the new variance?

Another question I have is: we initially decided not to integrate the tests in the `http_test_server` repository into CI in Vector as we expected that the tests in Vector would be sufficient to give us confidence. However, in this case, it wasn't until we attempted to run the `http_test_server` tests that we observed there could be issues. Do you think we need to expand the tests in Vector to cover more cases? Or are there other ways we could build confidence that ARC is functioning well and doesn't regress in the future?

bruceg committed:
Add support for a mean+variance running statistic, unused

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Make running mean stats math more accurate

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Merge remote-tracking branch 'origin/master' into arc-ewma-variance

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Drop unused MeanVar

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>
